tools/pyvmi/libvmi.py

```diff
@@ -0,0 +1,665 @@
+        status = lib.vmi_read_addr_ksym(self.vmi, symbol.encode(), value)
+        check(status)
+        return value[0]
+
+    def read_8_va(self, vaddr, pid):
+        value = ffi.new("uint8_t *")
+        status = lib.vmi_read_8_va(self.vmi, vaddr, pid, value)
+        check(status)
+        return value[0]
+
+    def read_16_va(self, vaddr, pid):
+        value = ffi.new("uint16_t *")
+        status = lib.vmi_read_16_va(self.vmi, vaddr, pid, value)
+        check(status)
+        return value[0]
+
+    def read_32_va(self, vaddr, pid):
+        value = ffi.new("uint32_t *")
+        status = lib.vmi_read_32_va(self.vmi, vaddr, pid, value)
+        check(status)
+        return value[0]
+
+    def read_64_va(self, vaddr, pid):
+        value = ffi.new("uint64_t *")
+        status = lib.vmi_read_64_va(self.vmi, vaddr, pid, value)
+        check(status)
+        return value[0]
+
+    def read_addr_va(self, vaddr, pid):
+        value = ffi.new("addr_t *")
+        status = lib.vmi_read_addr_va(self.vmi, vaddr, pid, value)
+        check(status)
+        return value[0]
+
+    def read_str_va(self, vaddr, pid):
+        value = lib.vmi_read_str_va(self.vmi, vaddr, pid)
+        if value == ffi.NULL:
+            raise LibvmiError('VMI_FAILURE')
+        return ffi.string(value).decode()
+
+    def read_unicode_str_va(self, vaddr, pid):
+        value = lib.vmi_read_unicode_str_va(self.vmi, vaddr, pid)
+        if value == ffi.NULL:
+            raise LibvmiError('VMI_FAILURE')
+        encoding = ffi.string(value.encoding).decode()
+        buffer = ffi.buffer(value.contents, value.length)[:]
+        self.free_unicode_str(value)
+        return buffer.decode(encoding)
+
+    # missing convert_str_encoding
+    def free_unicode_str(self, unicode_str):
+        lib.vmi_free_unicode_str(unicode_str)
+
+    def read_8_pa(self, paddr):
+        value = ffi.new("uint8_t *")
+        status = lib.vmi_read_8_pa(self.vmi, paddr, value)
+        check(status)
+        return value[0]
+
+    def read_16_pa(self, paddr):
+        value = ffi.new("uint16_t *")
+        status = lib.vmi_read_16_pa(self.vmi, paddr, value)
+        check(status)
+        return value[0]
+
+    def read_32_pa(self, paddr):
+        value = ffi.new("uint32_t *")
+        status = lib.vmi_read_32_pa(self.vmi, paddr, value)
+        check(status)
+        return value[0]
+
+    def read_64_pa(self, paddr):
+        value = ffi.new("uint64_t *")
+        status = lib.vmi_read_64_pa(self.vmi, paddr, value)
+        check(status)
+        return value[0]
+
+    def read_addr_pa(self, paddr):
+        value = ffi.new("addr_t *")
+        status = lib.vmi_read_addr_pa(self.vmi, paddr, value)
+        check(status)
+        return value[0]
+
+    def read_str_pa(self, paddr):
+        value = lib.vmi_read_str_pa(self.vmi,paddr)
+        if value == ffi.NULL:
+            raise LibvmiError('VMI_FAILURE')
+        return ffi.string(value).decode()
+
+    # write
```

Comment:
Redundant comment?

Reply:
it was just to separate all the read_* APIs from the write_* with a comment that should be view as a "section separator"